File: src/orchestrator/escalate.ts
import { ModelConfig } from '../types/config';

export interface EscalationState {
    currentModel: ModelConfig;
    attempts: number;
}

export interface EscalationPolicy {
    maxAttempts: number;
    ladder: ModelConfig[];
}

/**
 * Decide whether escalation is allowed.
 */
export function canEscalate(
    state: EscalationState,
    policy: EscalationPolicy
): boolean {
    if (state.attempts >= policy.maxAttempts) {
        return false;
    }

    return (
        indexOfModel(state.currentModel, policy.ladder) <
        policy.ladder.length - 1
    );
}

/**
 * Return the next model in the escalation ladder.
 */
export function escalateModel(
    state: EscalationState,
    policy: EscalationPolicy
): ModelConfig | null {
    if (!canEscalate(state, policy)) {
        return null;
    }

    const currentIndex = indexOfModel(
        state.currentModel,
        policy.ladder
    );

    return policy.ladder[currentIndex + 1];
}

/**
 * Find a model in the ladder by name.
 */
function indexOfModel(
    model: ModelConfig,
    ladder: ModelConfig[]
): number {
    const index = ladder.findIndex(
        (m) => m.name === model.name
    );

    if (index === -1) {
        throw new Error(
            `Model ${model.name} not found in escalation ladder`
        );
    }

    return index;
}

File: rust/rlm/src/orchestrator/escalate.rs
use crate::types::config::ModelConfig;

pub struct EscalationState {
    pub current_model: ModelConfig,
    pub attempts: u32,
}

#[derive(Debug, Clone, Default)]
pub struct EscalationPolicy {
    pub max_attempts: u32,
    pub ladder: Vec<ModelConfig>,
    pub max_model: String,
    pub fallback_models: Vec<String>,
}

pub fn can_escalate(state: &EscalationState, policy: &EscalationPolicy) -> bool {
    if state.attempts >= policy.max_attempts {
        return false;
    }

    index_of_model(&state.current_model, &policy.ladder) < policy.ladder.len() - 1
}

pub fn escalate_model(
    state: &EscalationState,
    policy: &EscalationPolicy,
) -> Option<ModelConfig> {
    if !can_escalate(state, policy) {
        return None;
    }

    let current_index = index_of_model(&state.current_model, &policy.ladder);
    policy.ladder.get(current_index + 1).cloned()
}

fn index_of_model(model: &ModelConfig, ladder: &[ModelConfig]) -> usize {
    ladder
        .iter()
        .position(|m| m.name == model.name)
        .unwrap_or_else(|| {
            panic!("Model {} not found in escalation ladder", model.name);
        })
}
